client/dashboard/me/preferences-language/index.tsx

```diff
@@ -0,0 +1,310 @@
+	const [ localData, setLocalData ] = useState< Partial< UserSettingsPreferences > | undefined >();
+	const mutation = useMutation( userSettingsPreferencesMutation() );
+	const [ savingData, setSavingData ] = useState<
+		Partial< UserSettingsPreferences > | undefined
+	>();
+
+	// After a successful save we reload the page with a query parameter and then
+	// display the success notice here, finally cleaning the URL so it stays the same.
+	useEffect( () => {
+		if ( typeof window === 'undefined' ) {
+			return;
+		}
+		const url = new URL( window.location.href );
+		if ( url.searchParams.has( 'settings-saved' ) ) {
+			createSuccessNotice( __( 'Language setting saved.' ), { type: 'snackbar' } );
+			url.searchParams.delete( 'settings-saved' );
+			window.history.replaceState( {}, '', url.toString() );
+		}
+	}, [ createSuccessNotice ] );
+
+	/**
+	 * When we save the language, in case we're using a locale_variant (a language without an official locale)
+	 * the API will return the parent language (example: es-cl will return 'es' in the 'language' field)
+	 * as such, we're overriding the language with the locale_variant, if present, this saves us from checking the data and allows us to
+	 * trust that the 'language' field contains a locale from the languages.
+	 */
+	if ( serverData?.locale_variant && serverData.locale_variant !== '' ) {
+		serverData.language = serverData.locale_variant;
+	}
+	const data = useMemo(
+		() => ( serverData ? { ...serverData, ...savingData, ...localData } : undefined ),
+		[ serverData, savingData, localData ]
+	);
+
+	if ( ! data ) {
+		return null;
+	}
+	// We need the CalypsoLanguage to show the % of translated content in the use_fallback_for_incomplete_languages
+	const selectedLanguage: CalypsoLanguage | undefined = data.language
+		? ( getLanguage( data.language ) as CalypsoLanguage )
+		: undefined;
+
+	const handleSubmit = ( e: React.FormEvent ) => {
+		e.preventDefault();
+		if ( ! localData ) {
+			return;
+		}
+		const mutationData = localData;
+		setLocalData( undefined );
+		setSavingData( mutationData );
+		mutation.mutate( mutationData, {
+			onSuccess: () => {
+				// Ensure the UI picks up the new language by reloading the page.
+				// Add a transient query param so we can show a success notice after reload.
+				const url = new URL( window.location.href );
+				url.searchParams.set( 'settings-saved', '1' );
+				// Replace to avoid adding an extra history entry.
+				window.location.replace( url.toString() );
+			},
+			onError: ( error ) => {
+				// Prepend previous attempted data back into local edits
+				setLocalData( ( current ) => ( { ...mutationData, ...current } ) );
+				createErrorNotice( error.message ?? __( 'Language setting could not be saved.' ), {
+					type: 'snackbar',
+				} );
+			},
+			onSettled: () => {
+				setSavingData( undefined );
+			},
+		} );
+	};
+
+	const isDefaultLanguageSelected = !! data.language && isDefaultLocale( data.language );
```

Comment:
```suggestion
	const isDefaultLanguageSelected = Boolean( data.language ) && isDefaultLocale( data.language );
```

Feels more readable to me - WDYT? But this may be because I came from VIP and VIP's codebase prefers `Boolean()`. Though if the codebase prefers `!!`, I'm good with that.

Reply:
@abdullah-kasim I like your approach more, but then TS would still require a check for that field existance since `isDefaultLocale` doesn't accept undefined (only `string` or `null`)

Reply:
```suggestion
	const isDefaultLanguageSelected = Boolean( data.language && isDefaultLocale( data.language ) );
```

Aw, that's unfortunate. I'd wrap everything up with `Boolean` in that case, or do `isDefaultLocale( data.language || '' )` - even though I might be living dangerously in that manner.

Still, just a nit for me since this doesn't break a linting rule 👍

Reply:
Yeah, for now I'll keep it this way as I've seen `!!` being used multiple times in the codebase.